Add `custom_modules` build option to compile external, user-defined C++ modules

Xrayez opened the pull request:
Closes #31457, namely as suggested by @fire in https://github.com/godotengine/godot/issues/31457#issuecomment-580605122, also @willnationsdev.
Closes godotengine/godot-proposals#565.

## Description

This PR adds ability to include external, user-defined C++ modules to be compiled as part of Godot:
```
scons platform=x11 tools=yes target=release_debug custom_modules="../project/modules"
```

Feature-wise this PR is similar to #36883 in a sense that both provide a command line option to compile something externally, with the following differences:
* only minimal core changes related to `doctool` bug, but mostly only build scripts are modified.
* doesn't introduce any new concepts, but makes existing build process more flexible for module development.
* can detect multiple modules at `custom_modules`, while #36883 aims to provide a way to compile a single C++ package (which can compile other modules though).
* full support for per-module custom documentation and editor icons.

One could just create `game` module and do this:

```python
src:
  - godot:
        - modules
  - project: # contains `project.godot`
        - modules:
            - game # regular C++ module
            .gdignore
```

Then run this command from within `godot` directory:
```
scons platform=x11 tools=yes target=release_debug custom_modules=../project/modules
```

with the added benefit of having ability to include any other module alongside.

The `register types` would look like this for a `game` module:

```cpp
#include "register_types.h"
#include "game.h"

void preregister_game_types() {

}
void register_game_types() {

}
void unregister_game_types() {

}
```

Pretty intuitive isn't it? 🙂 

## Features
* Can detect all available custom modules under `custom_modules` directory the same way as it does for built-in modules (not recursive).
* Multiple search paths can be specified similarly to `PATH` env var.
* Works with both relative and absolute paths on your filesystem.

## Benefits
* No need to manually **copy** module files directly to `godot/modules`, just point the path where your custom modules reside in your project.
* No need to [**symlink**](https://github.com/godotengine/godot/pull/28343#issuecomment-486215132) modules either to simulate the same behavior (I can finally remove all symlink hacks in my own project, which sometimes don't work on all platforms reliably).
* No need to [**fork** ](https://github.com/godot-extended-libraries/godot-tensorflow-godot)Godot, makes it user-friendly for those using Godot+modules to be versioned with `git` as submodules, as C++ modules can be independent from Godot repository.

## Implementation details
See commit description.

## Suggestions
Option name is subject to discussion, feedback welcome.

Commit:
Add `custom_modules` build option to compile external user modules

This patch adds ability to include external, user-defined C++ modules
to be compiled as part of Godot via `custom_modules` build option
which can be passed to `scons`.

```
scons platform=x11 tools=yes custom_modules="../project/modules"
```

Features:

- detects all available modules under `custom_modules` directory the
same way as it does for built-in modules (not recursive);
- works with both relative and absolute paths on the filesystem;
- multiple search paths can be specified as a comma-separated list.

Module custom documentation and editor icons collection and generation
process is adapted to work with absolute paths needed by such modules.

Also fixed doctool bug mixing absolute and relative paths respectively.

Implementation details:

- `env.module_list` is a dictionary now, which holds both module name as
  key and either a relative or absolute path to a module as a value.
- `methods.detect_modules` is run twice: once for built-in modules, and
  second for external modules, all combined later.
- `methods.detect_modules` was not doing what it says on the tin. It is
  split into `detect_modules` which collects a list of available modules
  and `write_modules` which generates `register_types` sources for each.
- whether a module is built-in or external is distinguished by relative
  or absolute paths respectively. `custom_modules` scons converter
  ensures that the path is absolute even if relative path is supplied,
  including expanding user paths and symbolic links.
- treats the parent directory as if it was Godot's base directory, so
  that there's no need to change include paths in cases where custom
  modules are included as dependencies in other modules.

Xrayez commented:
Managed to resolve remaining issues with collecting and generating custom **documentation** for external modules. I had to modify the doctool to distinguish between relative and absolute paths respectively.

Similarly fixed modules **editor icons** collection by editor icon builder.

All-in-all, this makes it work exactly like built-in modules, making this PR feature-complete IMO.

🤞 🤞

AndreaCatania commented:
This version is much better then the one that I did. Really good Job!

fire commented:
The flag `modules_search_path` is a bit long and wordy. Any suggestions?

I'm looking at the other examples like `target` or `bits` or `platform`.

Zireael07 commented:
--module(s)_path? (We can drop the s to shorten it even further?)

fire commented:
There is only one path to a folder of modules. So `module_path=~/Documents/game_project/modules/` sounds great.

Xrayez commented:
> --module(s)_path? (We can drop the s to shorten it even further?)

Multiple modules can be detected at provided path, so the option should suggest that, else it may lead to confusion that only a single module can be linked, which is not the case here.

> There is only one path to a folder of modules

See above. If the option supported multiple paths, the name would be `module_paths` otherwise. I'd personally go for `modules_path`, or even `modules_dir`.

Some more suggestions:

- `custom_modules` ([as described in docs](http://docs.godotengine.org/en/latest/development/cpp/custom_modules_in_cpp.html))
- `user_modules`

We could go further and introduce either `package` or `framework` keywords for that. 😛

Especially if the option could support multiple search paths. You could combine several module packages residing at different locations, but that's too much to ask I guess.

Xrayez commented:
Renamed `modules_search_path` to `custom_modules` to reduce verbosity. The rationale behind this:

* the name encodes necessary information that multiple modules must be passed (plural form).
* the help page is descriptive and the converter is smart enough to suggest you that directory path should be fed to detect custom modules in case of invalid input.
* ["Custom modules"](http://docs.godotengine.org/en/latest/development/cpp/custom_modules_in_cpp.html) comes from documentation, it can be added later to documentation page without any friction with existing terminology.

I've also added an explicit check to prevent passing a path to a **single** module (parent directory should be used instead in that case).

Xrayez commented:
I've took some time to test this further.

Fixed an issue with path includes. Namely when you include modules as dependencies of other modules:
```cpp
#ifndef YOUR_MODULE_H
#define YOUR_MODULE_H

#include "modules/my_module/my_class.h"

// ...
```

Now, `scons` treats the `custom_modules`'s parent directory as if it was Godot's base directory, so I advice putting your custom modules under `modules` directory (even if it's outside Godot) similarly to built-in modules (if you have module dependencies like this) so that include paths remain the same. But you can name it whatever you like of course, depending on whether you need to interact with built-in modules from within custom modules, and vice versa.

Xrayez commented:
To test this with existing project, I've had to port this PR to 3.2, so to whoever want to test this against 3.2, here's the commit: https://github.com/Xrayez/godot/commit/e237bf51f58e388a639e821927058ce1cabce93d.

Also providing a [raw link to 3.2 patch](https://github.com/Xrayez/godot/commit/e237bf51f58e388a639e821927058ce1cabce93d.patch), and some commands:

**Linux** (not tested)
```
curl https://github.com/Xrayez/godot/commit/e237bf51f58e388a639e821927058ce1cabce93d.patch | git apply -v
```

**Windows**
```
curl https://github.com/Xrayez/godot/commit/e237bf51f58e388a639e821927058ce1cabce93d.patch | Select-Object -Expand Content | git apply -v
```

Xrayez commented:
Here's an astonishing way to test this PR:

1. Move all (or some) built-in modules to whatever place on your filesystem under `modules` directory (except for `gdnative` module, it seems like it has core dependencies). Note: don't move an entire `godot/modules` directory, just individual modules.

2. Build. I'm using MSVC so it looks like this on my side:
```
scons platform=windows tools=yes target=release_debug bits=64 custom_modules="D:/tmp/modules/"
```

![godot-builtin-modules-as-custom](https://user-images.githubusercontent.com/17108460/76534975-dcbb2600-6482-11ea-8f73-94c6aff94739.png)

Compilation successful. Classes, editor icons and documentation are still there!

Notice absolute paths when compiling modules, that's the way `scons` knows about what modules are external.

Xrayez commented:
Added ability to specify multiple search paths for custom modules, not much changes internally.
```
scons platform=windows custom_modules="D:/common/editor/modules;../project/modules"
```
Notice that paths must be separated either by `;` or `:` (Windows/Unix respectively).

That way, you can define your C++ packages as a collection of modules, so to speak.

I haven't added support for checking `GODOT_CUSTOM_MODULES_PATH` environment variable or alike. This can be achieved by exporting/extending `SCONSFLAGS` instead:

**Unix:**
```
export SCONSFLAGS="custom_modules=../project/modules"
```
**Windows:**
```
$env:SCONSFLAGS = "custom_modules=../project/modules"
```

Also added a note that custom modules can override built-in ones at build-time. For instance, it's perfectly possible to maintain your own `gdscript` module within your project, and it will always shadow the built-in one. You can use this to allow something akin to local module fork, so to speak.

AndreaCatania commented:
@Xrayez I'm using this PR for some time now and it works really well. With this I'm not anymore constrained to have a specific godot repository per each proejct. Indeed, I just need to update my main godot source code once to be able to use it with all projects.

Also, the various compilations takes much much less time since the 90% of the code is already compiled.

But the most important thing is that now it's really simple handles the shared modules and I don't need anymore perform cherry-picks.

I didn't used docs and icons yet, but for me even without those two features is a really good job. Thanks for this contribution!

asheraryam commented:
I took a look at the code and it looks fine to me, the changes are straightforward and not invasive.

I would love for this to be merged soon, looking forward to it!

fire commented:
Is your 3.2 branch still updated?

Xrayez commented:
@fire I've updated `modules-search-path-3.2` branch with recent changes to this PR just now: https://github.com/Xrayez/godot/commit/22c0adc4afd5a6bf18e61d4ffe306e1ebd1e15e4 namely: https://github.com/godotengine/godot/pull/36922#issuecomment-599775440.

I don't think there's anything left to be added to this PR feature-wise. Depending on my own needs, I might keep updating the `3.2` version of this patch as well, but I hope that there won't be any merge conflicts to be resolved on the `3.2-stable` for that matter.

AndreaCatania commented:
Any idea when this can be merged or what is needed to start the process to merge this?

RaTcHeT302 commented:
> 
> 
> Any idea when this can be merged or what is needed to start the process to merge this?

i don't think c++ is getting enough love :(

Calinou commented:
@RaTcHeT302 If C++ wasn't getting any love, then nobody would be contributing to Godot :wink: 

It's just that pull request reviews are currently slowed down because the `master` branch is undergoing refactoring.

RaTcHeT302 commented:
> 
> 
> @RaTcHeT302 If C++ wasn't getting any love, then nobody would be contributing to Godot 😉
> 
> It's just that pull request reviews are currently slowed down because the `master` branch is undergoing refactoring.

i mean i would help if i knew how, but i'm not really familiar with the engine yet, and i haven't had the time to figure things out - my end goals would clash with godot either way, and i have a weird style, where i tend to over comment everything, as i figure things out along the way

fire commented:
Can you add the custom_modules flag to vsproj for Windows Visual Studio?

Xrayez commented:
@fire tbh I rarely use Visual Studio IDE (if at all), mostly VSCode, so not sure what needs to be made in order for this to work. I've read through the `Sconstruct` vsproj generation and there are some commands which can be modified I guess:

https://github.com/godotengine/godot/blob/6a0473bcc23c096ef9ee929632a209761c2668f6/methods.py#L504-L515

I suspect the `!option!` syntax is there to denote user-defined vars or so (?).

In any case this can be added as part of another PR I guess, lets review this one first. 🙂

Xrayez commented on SConstruct:
```diff
@@ -179,18 +177,43 @@ for k in platform_opts.keys():
+module_search_paths = ["modules"]  # Built-in path.
+
+if ARGUMENTS.get("custom_modules"):
+    paths = ARGUMENTS.get("custom_modules").split(os.pathsep)
```
```suggestion
    paths = ARGUMENTS.get("custom_modules").split(":")
```
I wonder whether it makes sense to rely on platform-specific path separator at all, yet that's not too critical.

Calinou replied:
If you want to specify an absolute path on Windows, you have to use `:` to specify the drive letter. This is why `;` should be used as a path separator there.

Xrayez replied:
Ah yes makes sense, it's just that I've seen Godot uses some double colon `::` in paths when running from command line, but this wouldn't be intuitive.

Xrayez commented:
I want to report that this also works from [within WSL with VSCode integration](https://code.visualstudio.com/docs/remote/wsl), I've just compiled `scons platform=javascript custom_modules=../project/modules` with no issues (the absolute paths resolve to something like  `/mnt/c/src/project/modules/...`).

akien-mga commented:
Thanks for bearing with me, I'm now ready to review this :)

I can also say +1 for the proposal/use case, I'll review the implementation more in-depth but a priori this should be good to integrate in the engine. Once merged in `master`, I'm also fine having this cherry-picked for `3.2` (ideally via a dedicated PR as the `black` formatting in `master` will make cherry-picking bothersome).

akien-mga commented on SConstruct:
```diff
@@ -179,18 +177,43 @@ for k in platform_opts.keys():
-    module_enabled = True
-    tmppath = "./modules/" + x
-    sys.path.insert(0, tmppath)
+# Detect modules.
+opts.Add("custom_modules", "Directory path containing custom modules to build along with built-in modules.", "")
```
Move this further up with other options (under "Advanced options" or "Components" I guess, as you prefer).

Xrayez replied:
Yeah this was my initial course of action, it's just that I've figured if we add module-specific options dynamically like `module_x_enabled`, it kinda made sense to move this option down there, but I think it's better to move this up for visibility then.

akien-mga commented on SConstruct:
```diff
@@ -179,18 +177,43 @@ for k in platform_opts.keys():
+module_search_paths = ["modules"]  # Built-in path.
+
+if ARGUMENTS.get("custom_modules"):
+    paths = ARGUMENTS.get("custom_modules").split(os.pathsep)
```
Wouldn't it be better to use `,`, which should be usable cross-platform? It makes both documentation and usage simpler if users don't need to differentiate between Windows and POSIX.

Xrayez replied:
If that makes it cross-platform, sure, but needs a mention that it's a comma-separated list of paths explicitly then, as you suggested below.

akien-mga commented on SConstruct:
```diff
@@ -179,18 +177,43 @@ for k in platform_opts.keys():
-    tmppath = "./modules/" + x
-    sys.path.insert(0, tmppath)
+# Detect modules.
+opts.Add("custom_modules", "Directory path containing custom modules to build along with built-in modules.", "")
```
Should mention that it can be a list of paths (comma-separated if we go with that, or `os.pathsep` separated otherwise, but as mentioned this makes documentation and usage trickier if users need to call `python -c "import os; print(os.pathsep)"` from their shell to have a cross-platform call).

akien-mga commented on SConstruct:
```diff
@@ -179,18 +177,43 @@ for k in platform_opts.keys():
-    sys.path.remove(tmppath)
+    try:
+        enabled = config.is_enabled()
+    except AttributeError:
+        pass
```
Won't this silently enable broken modules?

Xrayez replied:
Well for this line, the logic here is that if we don't define `config.is_enabled()`, we assume that a module is enabled by default (this was the default behavior before).

IMO the module is only broken when it can't compile.

akien-mga replied:
Ah indeed. I guess it's fine then.

akien-mga reviewed (approved):
Some nitpicks, but looks great overall.

Xrayez commented on SConstruct:
```diff
@@ -123,6 +121,7 @@ opts.Add(BoolVariable("use_precise_math_checks", "Math checks use very precise e
 opts.Add(BoolVariable("deprecated", "Enable deprecated features", True))
 opts.Add(BoolVariable("minizip", "Enable ZIP archive support using minizip", True))
 opts.Add(BoolVariable("xaudio2", "Enable the XAudio2 audio driver", False))
+opts.Add("custom_modules", "A list of comma-separated directory paths containing custom modules to build.", "")
```
I've shortened the description because the `black` formatter insisted on splitting this across 4 lines. 😛

akien-mga replied:
Yeah I've had this issue too, it's a pain :D

Xrayez commented on SConstruct:
```diff
@@ -181,18 +180,41 @@ for k in platform_opts.keys():
+# Detect modules.
+modules_detected = {}
+module_search_paths = ["modules"]  # Built-in path.
+
+if ARGUMENTS.get("custom_modules"):
+    paths = ARGUMENTS.get("custom_modules").split(",")
```
Would be nice to *actually* fetch `custom_modules` via `custom.py` as described in #38821, but none of my attempts worked:
```python
custom_modules = ARGUMENTS.get("custom_modules")
if not custom_modules:
    # Assuming the option is set via a file instead.
    custom_modules = opts.args.get("custom_modules", "") # Returns empty string, but defined in `custom.py`...
```

It kinda works if we update the construction environment with those options earlier, but that means calling `opts.Update(env_base)` twice. Then it can be changed to:

```python
custom_modules = env_base["custom_modules"]
```

Xrayez commented:
All in all I think this is ready to merge, I've just left some notes for some future changes if some more advanced features are needed.

akien-mga commented:
Thanks!

Xrayez commented:
> Once merged in master, I'm also fine having this cherry-picked for 3.2

Thanks, I'll work on the 3.2 version now then!
